From 4333ad96d5f2ded0579f03f6d495481fd1f79fe8 Mon Sep 17 00:00:00 2001
From: Jesper Brouer <brouer@redhat.com>
Date: Wed, 19 Mar 2014 11:29:54 -0400
Subject: [net] ipvs: fix AF assignment in ip_vs_conn_new()

Message-id: <20140319112954.13556.63359.stgit@dragon>
Patchwork-id: 78404
O-Subject: [RHEL7 net PATCH 1/2] ipvs: fix AF assignment in ip_vs_conn_new()
Bugzilla: 1074516
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Thomas Graf <tgraf@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Bugzilla: 1074516
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7224056

Backport of upstream commit: 2a971354e74f

 commit 2a971354e74f3837d14b9c8d7f7983b0c9c330e4
 Author: Michal Kubecek <mkubecek@suse.cz>
 Date:   Thu Jan 30 08:50:20 2014 +0100

    ipvs: fix AF assignment in ip_vs_conn_new()

    If a fwmark is passed to ip_vs_conn_new(), it is passed in
    vaddr, not daddr. Therefore we should set AF to AF_UNSPEC in
    vaddr assignment (like we do in ip_vs_ct_in_get()), otherwise we
    may copy only first 4 bytes of an IPv6 address into cp->daddr.

    Signed-off-by: Bogdano Arendartchuk <barendartchuk@suse.com>
    Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
    Acked-by: Julian Anastasov <ja@ssi.bg>
    Signed-off-by: Simon Horman <horms@verge.net.au>

Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/ipvs/ip_vs_conn.c b/net/netfilter/ipvs/ip_vs_conn.c
index a083bda..e884d17 100644
--- a/net/netfilter/ipvs/ip_vs_conn.c
+++ b/net/netfilter/ipvs/ip_vs_conn.c
@@ -871,11 +871,11 @@ ip_vs_conn_new(const struct ip_vs_conn_param *p,
  cp->protocol    = p->protocol;
  ip_vs_addr_set(p->af, &cp->caddr, p->caddr);
  cp->cport    = p->cport;
- ip_vs_addr_set(p->af, &cp->vaddr, p->vaddr);
- cp->vport    = p->vport;
- /* proto should only be IPPROTO_IP if d_addr is a fwmark */
+ /* proto should only be IPPROTO_IP if p->vaddr is a fwmark */
  ip_vs_addr_set(p->protocol == IPPROTO_IP ? AF_UNSPEC : p->af,
-         &cp->daddr, daddr);
+         &cp->vaddr, p->vaddr);
+ cp->vport    = p->vport;
+ ip_vs_addr_set(p->af, &cp->daddr, daddr);
  cp->dport          = dport;
  cp->flags    = flags;
  cp->fwmark         = fwmark;
-- 
1.7.1